Adding support for silent option to download

qyra opened the pull request:
Adding support for the silent=Bool keyword when calling download.
The default behaviour is to display progress indicators in wget and curl, to match the previous behaviour of download(). Passing silent=true will hide output from wget and curl.
This is primarly aimed at making download usable in async environments, without causing jumbled output.
The relevant documentation is also updated to note the new parameter, and that windows URLs must have the http:// suffix (wget / curl adds it automatically)

A previous pull request was made against the release-0.4 branch accidentally - see https://github.com/invenia/julia/pull/2 All comments from there should be addressed in this

tkelman commented on base/docs/helpdb/Base.jl:
```diff
@@ -130,13 +130,15 @@ Obtain the contents of an `IOBuffer` as an array, without copying. Afterwards, t
 
 """
-    download(url,[localfile])
+    download(url,[localfile]; silent=false)
```
needs rst signature update. and tests

tkelman commented:
x-ref a previous attempt at https://github.com/JuliaLang/julia/pull/12149, I think this is potentially a bit cleaner though

nalimilan commented on base/docs/helpdb/Base.jl:
```diff
@@ -130,13 +130,28 @@ Obtain the contents of an `IOBuffer` as an array, without copying. Afterwards, t
 
 """
-    download(url,[localfile])
+    download(url [localfile] [silent=false])
```
Why did you remove the comma? OTC, you need to add one before `silent`. Also, since there is `=false`, the brackets are not needed.

Same below.

nalimilan replied:
Ah, and since this is a keyword argument, you need to put `;` before `silent` (instead of a comma, contrary to what I said above).

akrasian replied:
I was having difficulty getting doc generation to work and trying to figure out why it was not matching the doc signatures. Editing the .rst files worked,I should not have pushed those changes. My current push corrects that.

nalimilan replied:
Looks like you didn't push the changes.

nalimilan commented on base/docs/helpdb/Base.jl:
```diff
@@ -130,13 +130,28 @@ Obtain the contents of an `IOBuffer` as an array, without copying. Afterwards, t
+    download(url [localfile] [silent=false])
 
 Download a file from the given url, optionally renaming it to the given local file name.
+Passing silent=true will suppress progress indicators.
```
Needs backquotes around `silent=true`.

nalimilan commented on base/docs/helpdb/Base.jl:
```diff
@@ -130,13 +130,28 @@ Obtain the contents of an `IOBuffer` as an array, without copying. Afterwards, t
+"""
+download
+
+"""
```
This is redundant with the docstring above, since `[localfile]` implies that argument can be omitted.

nalimilan commented on base/docs/helpdb/Base.jl:
```diff
@@ -130,13 +130,28 @@ Obtain the contents of an `IOBuffer` as an array, without copying. Afterwards, t
 or `fetch` to download the file and is provided for convenience. For production use or
 situations in which more options are needed, please use a package that provides the desired
-functionality instead.
+functionality instead. Platform specific details: In windows, download requires a full URL,
```
Please move `Platform specific details:` to a separate paragraph, as a title like this: `# Platform-Specific Details`. I would also rephrase it as something like "On Windows, a full URL must be provided, including the protocol prefix (e.g. "http://" for HTTP)."

akrasian commented:
Currently working on tests for this. There are a couple strange things to deal with because redirect_stdout() and redirect_stderr() don't actually let you redirect stdout and stderr from system calls for some reason.

akrasian commented:
Added a test file for interactiveutil which tests download silent/verbose, with or without a specified file, and addressed all comments made so far.

tkelman commented on test/interactiveutil.jl:
```diff
@@ -0,0 +1,110 @@
+
+"Runs an expression, and returns stdout, stderr, and the result."
+function capture_output(expression)
+	originalSTDOUT = STDOUT
```
4-space indent, no tabs

needs a rebase

tkelman commented on test/choosetests.jl:
```diff
@@ -33,7 +33,7 @@ function choosetests(choices = [])
         "enums", "cmdlineargs", "i18n", "workspace", "libdl", "int",
         "checked", "intset", "floatfuncs", "compile", "parallel", "inline",
-        "boundscheck", "error"
+        "boundscheck", "error", "interactiveutil"
```
will need to be added to the `net_required_for` array below

omus commented on test/interactiveutil.jl:
```diff
@@ -0,0 +1,110 @@
+	return (chop(outstr), chop(errstr), result)
+end
+
+port = 32744 #TODO - find generic way to get an unused port.
```
What you are looking for is [`listenany`](http://julia.readthedocs.org/en/latest/stdlib/io-network/#Base.listenany)

```julia
port, server = listenany(2000)
```

omus commented on test/interactiveutil.jl:
```diff
@@ -0,0 +1,110 @@
+	server_running = true
+	@async begin
+		server = listen(port)
+		while server_running
```
You need to declare `server_running` as global on [line 43](https://github.com/JuliaLang/julia/pull/15930/files#diff-125a8ef99a2617e382db614c9a77673fR43) and [line 56](https://github.com/JuliaLang/julia/pull/15930/files#diff-125a8ef99a2617e382db614c9a77673fR56). You can also delete [line 33](https://github.com/JuliaLang/julia/pull/15930/files#diff-125a8ef99a2617e382db614c9a77673fR33)

omus commented on test/interactiveutil.jl:
```diff
@@ -0,0 +1,110 @@
+
+downloader = :unknown
+for checkcmd in (:curl, :wget, :fetch)
+	if success(pipeline(`which $checkcmd`, DevNull))
```
`success` won't display STDOUT from the command. You can safely remove `pipeline` here.

tkelman replied:
which does not exist on windows unless someone happens to have cygwin or msys on their path (so appveyor may not exhibit a problem)

omus replied:
@tkelman that points out another issue. On Windows `download` uses a `ccall` which means we shouldn't be performing this check on Windows at all. Prepending `@unix_only` to this for loop should solve the issue.

omus commented on test/interactiveutil.jl:
```diff
@@ -0,0 +1,110 @@
+### Testing download(url,[localfile],[silent=false])
+
+"Runs an expression, and returns stdout, stderr, and the result."
+function capture_output(expression)
```
Could avoid using `eval` if you would change the signature take in a function and its arguments:
```julia
function capture_output(func::Function, args...; kwargs...)
    ...
    result = func(args...; kwargs...)
    ...
end

```

omus commented on test/interactiveutil.jl:
```diff
@@ -0,0 +1,110 @@
+	(errRead, errWrite) = redirect_stderr()
+	(outRead, outWrite) = redirect_stdout()
+
+	result = eval(expression)	
```
Should be in a `try` block where the finally block restores the original STDOUT/STDERR

omus commented on test/interactiveutil.jl:
```diff
@@ -0,0 +1,110 @@
+	errstr = join(map(x->Char(x), errors))
+	outstr = join(map(x->Char(x), output))	
+	# Remove the manually added \n from stdout and stderr
+	return (chop(outstr), chop(errstr), result)
```
Should be using [`chomp`](http://docs.julialang.org/en/latest/stdlib/strings/#Base.chomp) rather than [`chop`](http://docs.julialang.org/en/latest/stdlib/strings/#Base.chomp) here

omus commented on test/interactiveutil.jl:
```diff
@@ -0,0 +1,110 @@
+	redirect_stdout(originalSTDOUT)
+
+	errstr = join(map(x->Char(x), errors))
+	outstr = join(map(x->Char(x), output))	
```
Can be simplified to:
```julia
errstr = UTF8String(errors)
outstr = UTF8String(output)
```

omus commented on base/interactiveutil.jl:
```diff
@@ -466,9 +469,25 @@ end
+
+On windows, a full URL must be provided, including the protocol prefix (e.g "http://" for
+HTTP)
+"""
```
Documentation should be tied to the `download` function and not just this specific method.

omus commented on base/interactiveutil.jl:
```diff
@@ -466,9 +469,25 @@ end
+On windows, a full URL must be provided, including the protocol prefix (e.g "http://" for
+HTTP)
+"""
+function download(url::AbstractString; silent::Bool=false)
```
Documentation should be tied to the `download` function and not just this specific method.

omus commented on test/interactiveutil.jl:
```diff
@@ -0,0 +1,117 @@
+if downloader in (:curl, :wget)
+    @test length(errstr) == 0
+    @test length(outstr) == 0
+end
```
Probably best if we just test that their is no output for all downloader since Windows and `fetch` are both supposed to be silent.

omus commented on test/interactiveutil.jl:
```diff
@@ -0,0 +1,117 @@
+    # Progress appears on stderr for curl and wget
+    @test length(errstr) > 50
+    @test length(outstr) == 0
+end
```
Write an `else` test that expects there to be no output. This would cover both Windows and `fetch`

omus commented on test/interactiveutil.jl:
```diff
@@ -0,0 +1,117 @@
+html = "<html>success</html>"
+
+function start_server()
+    global server_running, port, html
```
`html` doesn't need to be global. Maybe just pass it in as an argument to `start_server`?

omus commented on test/interactiveutil.jl:
```diff
@@ -0,0 +1,118 @@
+        # Immediately restore output so errors are printed.
+        redirect_stderr(originalSTDERR)
+        redirect_stdout(originalSTDOUT)
+        println("Error checking output of $func: args=$args, kwargs=$kwargs - $exception")
```
Probably should still throw an exception. I would just do `rethrow()`

Commit:
Adding support for silent option to download. Also updates several out of date .rst files on master.

Updating docs

Regenerating RSTs for documentation

Adding automated tests for download function in interactiveutil

Server for tests will stop once test completes

Correcting function signature in docs, silent should be after ;

Renamed capture_output to check_output, matching the similar python functions name, using a try block to avoid issues while calling passed function, and autodetecting an unused port

Docs applied to all download functions, checkout output prints error messages, tests added to net_required_for, removing DevNull pipe from success

Enforcing 0 length stdout with fetch and windows download

Passing html to start_server rather than using it as a global

Updating comment on check_output

Rethrowing exceptions from test

Disabling windows test, until `@async` issue is resolved

akrasian commented:
Fixed up the conflicts with master. Running `make julia-genstdlib` created some changes in docs/stdlib/parallel.rst and docs/stdlib/base.rst which were not on the master, so apparently a commit made it in without the .rst files being regenerated. I just reset those files to the master version, ignoring the newly generated changes, since they had nothing to do with what I was working on in order to avoid any conflict.

omus commented on test/interactiveutil.jl:
```diff
@@ -0,0 +1,119 @@
+start_server(html)
+
+# Test verbose download, no specified file
+outstr, errstr, result = check_output(download, "localhost:$port")
```
Windows tests are probably failing since this is missing `http://` prefix

akrasian replied:
That, and @async doesn't appear to work in windows. The code in the async block never runs, even if it's just a print statement. I'll find some other way to test in windows / disable the test in windows soon.

omus commented:
@akrasian be careful to use backticks around macro names in git commit messages as it causes mentions to happen on GitHub.
